### PR TITLE
New feature: web hooks

### PR DESCRIPTION
**Pull request by lukaszlenart**
This PR adds new functionality to Codebrag allowing to define hooks on different internal events which are then propagated to external listeners (hosts).
 - [x] hooks
 - [x] serializers
 - [x] documentation

### CODE REVIEW

**Comment by JakeShirley** [comment]
This sounds awesome! Is this going to get merged?

**Comment by lukaszlenart** [comment]
This is ready for review, working on documentation now

**Comment by lukaszlenart** [comment]
The PR is ready, all the tasks accomplished :)

File: codebrag-service/src/main/scala/com/softwaremill/codebrag/eventstream/EventHookPropagator.scala [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      }
+      notifyListeners(NewCommitsLoadedHook(user, event.repoName, event.currentSHA, event.newCommits, event.hookName))
+
+    case (event: NewUserRegistered) =>
+      val user = userDao.findById(event.userId)
+      notifyListeners(NewUserRegisteredHook(user, event.login, event.fullName, event.hookName))
+
+  }
+
+  private def notifyListeners(hook: Hook) = {
```
Comment (adamw) [inline codebrag-service/src/main/scala/com/softwaremill/codebrag/eventstream/EventHookPropagator.scala]: Hooks are always constructed, even if there are no listeners. Given that this involves DB operations, might be a bit wasteful?

File: codebrag-service/src/main/scala/com/softwaremill/codebrag/eventstream/InternalSerializers.scala [hunk 2]
```diff
@@ -0,0 +1,61 @@
+package com.softwaremill.codebrag.eventstream
+
+import com.softwaremill.codebrag.domain.reactions.{UnlikeEvent, LikeEvent}
+import com.softwaremill.codebrag.domain._
+import org.json4s.ext.JodaTimeSerializers
+import org.json4s.{DefaultFormats, FieldSerializer}
+import org.json4s.FieldSerializer._
+
+object InternalSerializers {
```
Comment (adamw) [inline codebrag-service/src/main/scala/com/softwaremill/codebrag/eventstream/InternalSerializers.scala]: Aren't these only used by hooks?

File: codebrag-service/src/main/scala/com/softwaremill/codebrag/eventstream/EventHookPropagator.scala [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import com.softwaremill.codebrag.common.Clock
+import com.softwaremill.codebrag.dao.commitinfo.CommitInfoDAO
+import com.softwaremill.codebrag.dao.events.NewUserRegistered
+import com.softwaremill.codebrag.dao.user.UserDAO
+import com.softwaremill.codebrag.domain._
+import com.softwaremill.codebrag.domain.reactions._
+
+import akka.actor.Actor
+import com.ning.http.client.StringPart
+import scala.concurrent.ExecutionContext.Implicits.global
```
Comment (adamw) [inline codebrag-service/src/main/scala/com/softwaremill/codebrag/eventstream/EventHookPropagator.scala]: Use the execution context from the actor's dispatcher?

File: codebrag-service/src/main/scala/com/softwaremill/codebrag/eventstream/EventHookPropagator.scala [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+import com.softwaremill.codebrag.common.Clock
+import com.softwaremill.codebrag.dao.commitinfo.CommitInfoDAO
+import com.softwaremill.codebrag.dao.events.NewUserRegistered
+import com.softwaremill.codebrag.dao.user.UserDAO
+import com.softwaremill.codebrag.domain._
+import com.softwaremill.codebrag.domain.reactions._
+
+import akka.actor.Actor
+import com.ning.http.client.StringPart
```
Comment (adamw) [inline codebrag-service/src/main/scala/com/softwaremill/codebrag/eventstream/EventHookPropagator.scala]: unused?

File: codebrag-service/src/main/scala/com/softwaremill/codebrag/eventstream/InternalSerializers.scala (hunk 2 above)
Comment (lukaszlenart) [inline codebrag-service/src/main/scala/com/softwaremill/codebrag/eventstream/InternalSerializers.scala]: Yes, but those serializers define how to serialize events or entities into JSON not hooks., so they can be used with any other future JSON serialization.

File: codebrag-service/src/main/scala/com/softwaremill/codebrag/eventstream/EventHookPropagator.scala (hunk 1 above)
Comment (lukaszlenart) [inline codebrag-service/src/main/scala/com/softwaremill/codebrag/eventstream/EventHookPropagator.scala]: :+1: defined a guard

**Comment by lukaszlenart** [comment]
@adamw done, ready for another shot!

**Comment by adamw** [comment]
LGTM :)

**Pull request merged**